### ISSUE TITLE
`NaN` equivalent for `astropy.time.Time`

### ISSUE THREAD

**Issue by MaxNoe** [astropy/astropy#6509]
Is there an equivalent of a `NaN` for `astropy.time.Time`? Like pandas `NaT` (Not a Time)?

**Comment by pllim** [comment]
`not isinstance(var, astropy.time.Time)`?

**Comment by MaxNoe** [comment]
Wat? I think you are missing the point.

It should be an instance of astropy.time.Time! Just like `np.nan` is an instance of `float`.
An instance conveying the meaning that the time is not valid.

**Comment by pllim** [comment]
Not a direct answer to your question but currently `NaT` becomes a masked entry in `Table` (you probably already knows this but I am stating it here for others' future reference):
```python
>>> import pandas as pd
>>> from astropy.table import Table
>>> s = pd.to_datetime(pd.Series(['Jul 31, 2009', '2010-01-10', None]))
>>> df = pd.DataFrame(data={'time': s})
>>> df
        time
0 2009-07-31
1 2010-01-10
2        NaT
>>> t = Table.from_pandas(df)
>>> t
<Table masked=True length=3>
             time            
        datetime64[ns]       
-----------------------------
2009-07-31T00:00:00.000000000
2010-01-10T00:00:00.000000000
                           --
```
p.s. `Series` values taken from https://pandas.pydata.org/pandas-docs/stable/timeseries.html

**Comment by taldcroft** [comment]
@MaxNoe - there is no `NaT`-equivalent in astropy `Time` at this moment, but there is a mostly-complete PR #6028 which adds support.  This allows for masked Time arrays much like `MaskedColumn` in `Table` or `MaskedArray` in numpy.  It also lets you do all the table operations like join or stacking that can generate missing values.

So unfortunately there is no good news for the current 2.0 or even master, but this should definitely be in the 3.0 release at the end of this year.

**Comment by MaxNoe** [comment]
@taldcroft While the PR you mentioned enables support for masking array time values, there is still now support for a scalar time value that signals missing data or invalid values.

We just ran into this again. Why can we not simply allow `np.nan` as value for the time attributes?
Why explicitly fail?

Missing values occur all the time (no pun intended) and everywhere in the real world. Prohibiting this seems very unreasonable to me.

**Comment by taldcroft** [comment]
@MaxNoe - can you give some examples of how you would like to use this?  I'm not entirely clear on your use case so examples would help.

**Comment by MaxNoe** [comment]
We have hardware failures of a certain telescope component sometimes that results in missing timestamps for certain events.

The usecase is simply being able to mark missing/invalid data. 

`numpy` has `np.NaT` for this in its `datetime64` datatype.

```
In [1]: import numpy as np                                                                                                                    

In [2]: np.array([np.nan]).astype('datetime64[s]')                                                                                            
Out[2]: array(['NaT'], dtype='datetime64[s]')
```

**Comment by taldcroft** [comment]
Your example just shows arrays, which are supported now and can give (I think) the same functionality you just showed.
```
In [5]: t = Time([1, 2], format='cxcsec')                                       

In [6]: t[0] = np.nan                                                           

In [7]: t.mask                                                                  
Out[7]: array([ True, False])
```
But you said something about a scalar Time object that is effectively "missing".  I want to know how you want to use that.

**Comment by taldcroft** [comment]
At the simplest level do you want to do `Time(np.nan)`?  What other operations would be useful?

**Comment by MaxNoe** [comment]
> At the simplest level do you want to do Time(np.nan)?
Yes. 

>  What other operations would be useful?
Normal nan propagation. E.g.:

*  `Time(np.nan) + 5 * u.s` ⇒ `Time(np.nan)`
* `Time.now() + np.nan * u.s` ⇒ `Time(np.nan)` 

The second case errors right now, which is why  this came up again yesterday.

In my example, the second case is actually a scalar (`np.array(np.nan)` is a scaler).

With your `nan` array example, there is a pretty big problem, as there is no scaler `nan` Time value,
this happens when accessing the value set to nan:

```python
In [1]: from astropy.time import Time
In [2]: import numpy as np
In [3]: t = Time([1, 2], format='unix')
In [4]: t.format = 'iso'
In [5]: t
Out[5]: <Time object: scale='utc' format='iso' value=['1970-01-01 00:00:01.000' '1970-01-01 00:00:02.000']>
In [6]: t[1] = np.nan
In [7]: t
Out[7]: <Time object: scale='utc' format='iso' value=['1970-01-01 00:00:01.000' --]>
In [8]: t[1]
Out[8]: <Time object: scale='utc' format='iso' value=1970-01-01 12:00:00.000>
```

That deserves a *wat*, don't you think?

**Comment by MaxNoe** [comment]
So what exactly is the motivation to allow setting an element of an array `Time` to `nan` but not a single `Time`.

That is highly confusing to me.

**Comment by MaxNoe** [comment]
Also, the error one get's for the second case is very unhelpful:

```
In [20]: Time.now() + np.nan * u.s
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/.local/anaconda3/lib/python3.7/site-packages/astropy/units/quantity_helper/converters.py in converters_and_unit(function, method, *args)
    176             try:
--> 177                 if can_have_arbitrary_unit(maybe_arbitrary_arg):
    178                     converters = [None, None]

~/.local/anaconda3/lib/python3.7/site-packages/astropy/units/quantity_helper/converters.py in can_have_arbitrary_unit(value)
    121     """
--> 122     return np.all(np.logical_or(np.equal(value, 0.), ~np.isfinite(value)))
    123

TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-20-285622984270> in <module>
----> 1 Time.now() + np.nan * u.s

~/.local/anaconda3/lib/python3.7/site-packages/astropy/units/quantity.py in __array_ufunc__(self, function, method, *inputs, **kwargs)
    442         # consistent units between two inputs (e.g., in np.add) --
    443         # and the unit of the result (or tuple of units for nout > 1).
--> 444         converters, unit = converters_and_unit(function, method, *inputs)
    445
    446         out = kwargs.get('out', None)

~/.local/anaconda3/lib/python3.7/site-packages/astropy/units/quantity_helper/converters.py in converters_and_unit(function, method, *args)
    190                                 .format(function.__name__,
    191                                         args[0].__class__.__name__,
--> 192                                         args[1].__class__.__name__))
    193
    194         # In the case of np.power and np.float_power, the unit itself needs to

TypeError: Unsupported operand type(s) for ufunc add: 'Time' and 'Quantity'
```

**Comment by mhvk** [comment]
@MaxNoe - arriving here via #10272, and thinking we should keep the general discussion here. I like the idea of being able to handle `nan` graciously. One possibility is to make the time immediately masked (which is what is currently done by setting `jd2` to `nan`, and the effects of which you see already a bit above, though also with some problematic examples). 

The other would be to have an unmasked `nan` (say in `jd1`). An issue with this is that it will cause complaints any time one does, say, a scale transformation and thus call an `erfa` routine. I'm a little wary of adding more checks around it, unless one can be sure it doesn't slow down regular stuff.

**Comment by mhvk** [comment]
Now having commented on #10272, I think first really should be to see if there is a need to distinguish masked from `isnan` (in my comment above, I'm already discussing implementation too much - the real question is one of what is needed to be represented; we should for now assume we can get masking to do whatever we want).

**Comment by Cadair** [comment]
I have a concrete use for real actual nan. I have implemented Time support in gWCS in the APE 14 framework. When doing transforms out of bounds for the transform, the default behaviour for all the other coordinate types other than Time we return a NaN for that element of the array. For time I have had to special case it and mask the output. The issue however is I don't think that the downstream consumers of the output are going to handle the mask correctly.

**Comment by maxnoe** [comment]
Yes, this is very similar to our use case, for which I initially opened this issue.

I couldn't directly make it work and had to focus on other things for the time being, but the PR I opened contains test cases how I envision a scalar astropy invalid `Time` should work: https://github.com/astropy/astropy/pull/10272

**Comment by maxnoe** [comment]
I think the main problem is that it requires implementing `__array_ufunc__` for `Time` and that is not simply possible for the one use-case but than has to directly implement all currently otherwise supported ufuncs in there using the new numpy API for that. 

Which is something I did not manage. @mhvk and I thought about starting a branch where we try to join forces to get the `__array_ufunc__` for time working, but this has not happend yet for the same reasons.

**Comment by taldcroft** [comment]
@Cadair @maxnoe - I like the idea of defining the desired behavior through tests or examples. 

@cadair - have you looked at those tests in #10272, and does that match what you are looking for? In particular I'm not sure if you consider returning `True` for an `np.isnan(tm)` to be the same as "we return a NaN for that element of the array".

> the default behaviour for all the other coordinate types other than Time we return a NaN for that element of the array.

Can you elaborate on what are the "other coordinate types" that return NaN? I'm also trying to understand if this fits into the concept of missing or masked data, which Astropy still doesn't support for Quantity and coordinates. @mhvk and I have had a (friendly) years-long disagreement about how to implement missing data in Quantity. I have advocated using NaN, and @mhvk wants to use a separate mask.

**Comment by maxnoe** [comment]
NaN is not usable for missing data. This was a major design flaw in pandas that was rectified with version 1.0. Please don't repeat mistakes that others have already made for us.

**Comment by taldcroft** [comment]
My understanding is that Pandas still uses `NaN` for float types. The design flaw mostly related to int and string types. It isn't obvious to me how they can replace this for floats and still maintain performance, but I could be missing something. The point is that they are using sentinel values in the array instead of carrying a separate mask array (AFAIK, I'm not a pandas expert).

**Comment by taldcroft** [comment]
Does anybody have a top-level understanding of how the pandas nullable int type is implemented?  I'm curious. I just did a quick test and it seems to be about 50% slower than the standard pandas integer for a simple addition test.

**Comment by maxnoe** [comment]
Quantities can have integer dtype.

**Comment by taldcroft** [comment]
> Quantities can have integer dtype.

Indeed, but in practice I suspect they represent a very tiny fraction of the use of Quantity in the wild. Given that `[1, 2] * u.m` gives a float dtype, integer Quantity is definitely a second-class citizen. I think nobody would complain if we announced that masked Quantities were available but only for floats. 

It has been 3.5 years since I opened #6031, and in that time we have been without support for missing values in Quantity. From the Table perspective this makes me unhappy.  Perfect is the enemy of good, and it would be very *good* to have some kind of masked quantity well before another 3.5 years passes. Certainly improvements in numpy have made the job of rewriting MaskedArray easier, but even with the very deep respect I have for @mhvk, we're still waiting... :smile:

**Comment by mhvk** [comment]
The problem with a sentinel value is mostly that what should be simple operations, like finding the minimum or mean, become slow because `nan` has to be special-cased. With a `Masked` class, that penalty only is paid when one actually cares about masked values.

Note that I actually did produce a `Masked` class that works well for `Quantity` - I got stuck somewhat because I hoped it also would work for `MaskedColumn`, but then one has to start to reproduce all the quirks that `MaskedArray` has, so it becomes a question to what extent one allows some minor deviations.

Looking back, I see I pushed this as a draft last June (see #10423) but have not had any comments. Please have a look, e.g., at the table test cases to see whether things look OK (mostly it makes changes to make sure we don't just assume `MaskedColumn` is a subclass of `MaskedArray`). IIRC, masked `Quantity` (and subclasses) actually worked quite well (lots of test cases...).

I can rebase if there is real interest...

**Comment by taldcroft** [comment]
@mhvk - sorry, somehow I missed that!  I'll have a look!

**Comment by mhvk** [comment]
The current inability to do `time_delta < nan-quantity` (#15230) reminded me of this issue. It has been a long time, and in the meantime we do have support for masked times, but those use `np.nan` as a signal for a mask, in a way preventing its use for indicating that there is an entry, but that it is "Not a Time" (NaT for numpy/datetime).

A possible way forward is 
1. Move `Time` to use the new `Masked` class underneath, which removes the need for the `np.nan` sentinel in `jd2`;
2. Use the sentinel instead to indicate NaT.

Since the sentinel is never exposed, this is in principle doable. It certainly made me revisit my earlier attempt at (1) - see #15231

**Comment by maxnoe** [comment]
@Cadair things for bringing this up again, I think the old PR that was closed due to inactivity is still a good starting point, it only adds test cases for the examples discussed here:

https://github.com/astropy/astropy/pull/10272/files

**Comment by mhvk** [comment]
Note that `Time` now does used a `Masked` class, so it is free to use `np.nan` to indicate NaT. Hence, I think it should now be reasonably straightforward (famous last words??).

**Comment by maxnoe** [comment]
@taldcroft I wouldn't consider this the correct behavior after the comment by @mhvk here https://github.com/astropy/astropy/pull/16713#discussion_r1687908315

Because if being NaT and being Masked are two different, separate states, then setting an element to nan should result in that element being NaT, not being masked.

So I think the correct behavior would be:


```
In [2]: t = Time([1, 2], format="cxcsec")
In [3]: t[0] = np.nan
In [4]: t
Out[4]: <Time object: scale='tt' format='cxcsec' value=[NaT, 2.]>
```

and:

```
In [2]: t = Time([1, 2], format="cxcsec")
In [3]: t.mask[0] = True
In [4]: t
Out[4]: <Time object: scale='tt' format='cxcsec' value=[--, 2.]>
```

**Comment by taldcroft** [comment]
> I wouldn't consider this the correct behavior

I think it is expected behavior for the current code which only has a concept of masked values. Setting an element to `np.nan` is currently a proxy for setting the mask to True.

Your desired behavior of `np.nan` => `NaT` seems reasonable at face value, but now I'm wondering about API and code breakage if we do that. Hmm.

**Comment by mhvk** [comment]
I don't know that there is a huge worry about API with setting with `np.nan`, but one could envisage `np.nan` being equal to NaT with mask set? 
Alternatively, we could simply recognize `np.datetime64("NaT")` (and perhaps provide an instance of it on `astropy.time` -- I actually thought numpy had this, but I cannot seem to find it! there only seems to be `np.isnat`)

**Comment by taldcroft** [comment]
I just checked and at least the behavior with `np.nan` => masked is not in the documentation. Instead users should be using `np.ma.masked` as the sentinal value in that case. So maybe the API / breakage issue is not so bad.
https://docs.astropy.org/en/stable/time/index.html#missing-values

I also went looking for `np.nat` and was disappointed. FWIW there is `pandas.NaT`. If we are looking for some special value, would the string `"NaT"` work? 

This is already implied in one of the tests which does ` t = Time("NaT", format=fmt)`. From that it seems natural that `tm[0] = "NaT"` would do the same thing for an array-valued time `tm`.

**Comment by mhvk** [comment]
Yes, makes sense to allow "NaT" also for non-string valued time formats.

**Comment by taldcroft** [comment]
One question that is not obvious for me is the representation of `NaT` for various time formats. Consider that a format type can be float, str, int, object. In fact the answer is `Any`.  So what comes out for `NaT` values when you do `tm.value`?

**Comment by maxnoe** [comment]
Could be solved by adding a `NaT` class variable on the format?

**Comment by taldcroft** [comment]
Looking through this issue, I'm starting to feel like we are conflating two issues:
1. Being able to use `np.nan` in some contexts to initialize a `Time` with missing values or do time arithmetic.
2. Adding a brand new concept of `NaT`.

I think that (1) can be done by working with the existing masked time machinery and improving some interface code (without major changes).

I am wondering if (2) is really required to address the use cases put forward so far. For instance the use-case put forward by @cadair to have time values `tm.value` output `np.nan` for missing values is problematic if the time format is not float type.

Pandas `NaT` is explicitly a marker for "missing values" (https://pandas.pydata.org/docs/reference/missing_value.html). Astropy is consistently using masks for missing values to avoid issues with representation and how to store sentinel values for different data types.

Adding a `NaT` to astropy time would then seem to be adding a second way of indicating missing values and therefore cause confusion, not to mention much more code complexity. I was really happy when we could remove all the code that was stuffing `np.nan` into `jd2`. (I never liked that code I wrote but it did fill a need at the time.)

**Comment by maxnoe** [comment]
> Consider this:

This is exactly the case I was considering...

> You cannot (or don't want to) put any NaT object into the output array since that would force turning into object type.
> Astropy time is different from Pandas and numpy in having a very flexible format attribute,

This is exactly the reason why I proposed a *class attribute* on the *format* class.

> Could be solved by adding a NaT class variable on the format?

Something like this:

```
class TimeFormatNumeric(TimeFormatBase):
    nat_value = np.nan

class TimeFormatString(TimeFormatBase):
    nat_value = "NaT"
```

And then:

```
>>> t = Time(np.nan, format="mjd")
>>> t.value
nan
>>> t.isot
"NaT"
```

**Comment by taldcroft** [comment]
@maxnoe - fair enough, I didn't get that you were implying a per-format class attribute. That can work, and presumably the test for NaT becomes something like `t.value == t.nat_value`. 

Strictly speaking this does not work for @Cadair's use-case since the output NaT marker depends on the format and does not have to be NaN:
> I have a concrete use for real actual nan. I have implemented Time support in gWCS in the APE 14 framework. When doing transforms out of bounds for the transform, the default behaviour for all the other coordinate types other than Time we return a NaN for that element of the array. 

@maxnoe - earlier you provided the fundamental driver for this:
> The usecase is simply being able to mark missing/invalid data.

That is already provided by a robust masking capability, so at this point it seems like the goal is a SECOND way of doing that which uses a format-class-specific sentinel in the `value` attribute instead of returning a `Masked` array. But from a conceptual and computational standpoint (e.g. `np.mean(tm)`), `NaT` would be exactly the same as masking.

Is that a correct characterization?

**Comment by mhvk** [comment]
I am also slightly worried about the actual use case. In the end, with masking, one can always do `time.<format>.filled(np.nan)` (though then it is up to the user to ensure `<format>` produces a float).

**Comment by maxnoe** [comment]
#12434 is now implemented, and also works for scalar values, as expected I would say:

```
In [18]: t = Time(Masked(0, mask=True), format='mjd')

In [19]: t
Out[19]: <Time object: scale='utc' format='mjd' value=    ———>

In [20]: t.filled(Time(0, format="mjd"))
Out[20]: <Time object: scale='utc' format='mjd' value=0.0>

In [21]: t.filled(Time(0, format="unix"))
Out[21]: <Time object: scale='utc' format='mjd' value=40587.0>
```

I think this should now make it possible to support `np.isnat` on `Time` objects.

**Comment by taldcroft** [comment]
I took a quick refresh of the long previous thread. I'm not sure I got it all again, but I did note that `t = Time(np.nan, format="mjd")` now fails (as part of now using `Masked` under the hood).

> I think this should now make it possible to support np.isnat on Time objects.

Would be always equivalent to `tm.mask`? I can see the benefit of supporting the `np.isnat` pathway for consistency with other time-like objects.

**Comment by maxnoe** [comment]
>  but I did note that t = Time(np.nan, format="mjd")

That's a bit surprising to me, why not accept the nan and mask it?

**Comment by maxnoe** [comment]
This is also ... surprising:

```
In [16]: t = Time(Masked(np.nan, mask=True), format="mjd")

In [17]: t.unmasked.iso
Out[17]: '2000-01-01 12:00:00.000'

In [18]: t.mjd.unmasked
Out[18]: array(51544.5)
```

**Comment by mhvk** [comment]
@maxnoe - the erfa functions cannot deal with `np.nan`, which is why on construction of a `Time` object any `nan` get replaced with J2000 by a call to `_fill_masked_values`. If we want to preserve `nan` to convert them to `NaT` later, we could repurpose @taldcroft's original scheme where he used `jd2=nan` to indicate masked values to know indicate `NaT`. But it means that on every call into an `erfa` routine we'd have to replace `nan` with `0`. I think this doesn't have to affect performance too badly if we cache whether or not any `nan` is present at all. But sadly it means it is not quite trivial to implement either.

p.s. An alternative might be to store `nan`, `inf`, and `ninf` as bits in the mask array, treating it as `uint8` - unless special care is taken, though, that would mean every one of those would be seen as masked, which may or may not be good. So, probably too much hackery...

**Comment by taldcroft** [comment]
I just discovered that `np.isnat` is a ufunc, not a numpy function. 

@mvhk knows all this of course. I just asked Claude Fable to have a go at supporting `np.isnat` for Time but adding `__array_ufunc__` broke a whole bunch of stuff (initially 28 unrelated failing tests). Claude was able to whittle this down to just 2 failures, but it was looking ugly. Maybe a human expert could do better, but I now have a better understanding of what is going on:
- `__array_function__` is a safe, additive hook — nothing in the language dispatches through it, and a class that doesn't implement a given function just doesn't support it.
- `__array_ufunc__` is the opposite: defining it puts your class in the middle of every arithmetic and comparison expression it appears in.

**Comment by maxnoe** [comment]
@taldcroft Yes, I started to had a go at this in https://github.com/astropy/astropy/pull/10272, but closed it as I could not get it right at the time

**Comment by taldcroft** [comment]
About the `np.nan` initialization, I have a vague recall of purposely wanting to move away from `np.nan` as a sentinel or proxy for missing values. Again this relates to NaN not being universally apropos as a missing value in a Time object. I will say that `np.ma.masked` would not suffer that problem, but it is explicitly excluded as well for reasons I don't recall:
```
File ~/git/astropy/astropy/utils/masked/core.py:92, in get_data_and_mask(array)
     86         raise AttributeError(
     87             f"'{type(array).__name__}' object has a 'mask' attribute but not an "
     88             "'unmasked' attribute (and is not an np.ma.MaskedArray instance)."
     89         ) from None
     91 if array is np.ma.masked:
---> 92     raise ValueError("cannot handle np.ma.masked.")
```

**Comment by mhvk** [comment]
Note that `np.ma.masked` is just a masked 0; the reason that it nevertheless is explicitly set to fail in `get_data_and_mask` is that conceptually it carries no `data`, so what should the function return?

**Comment by mhvk** [comment]
On `__array_function__` vs `__array_ufunc__` - the former is easier in part because `Time` already carries it from `ShapedLikeNDArray`, which does the proper forwarding for any unrecognized function. The same can be done for `__array_ufunc__` but as one starts from scratch, it is messier. Anyway, I'm fairly sure I could get it to work, but I think we need a plan on how to keep track of `nan` beforehand, and decide if that's worth it. In principle, the `jd2=nan` route seems easiest (could even extent to infinities).

**Comment by taldcroft** [comment]
I'm very much not enthusiastic about returning to injecting `jd2=nan` within the internals of `TimeFormat`. It substantially increases code complexity by having two ways of representing missing values. In fact the `nan` case might introduce ambiguity about whether `nan` is missing or just `nan` and the user can interpret it however they want. 

If I understand, the intent of "preserving nan" would be something like:
```
>>> tm = Time(np.nan, format="jd")
>>> tm.value
np.nan  # True NaN
>>> tm.iso
---  # ?? but probably now missing.
```

Allowing for `np.nan` on input for numerical formats seems possible in principle, where it would convert to a masked value just like providing `Masked(0, masked=True)` (which replaces the data with a fixed value that keeps ERFA happy).

Overall I would prefer to support only `np.ma.masked` as the sentinel since it is not numeric (conceptually). Then there is just one way to do it, and inputs like `Time(["2000:001", np.ma.masked])` would make sense. Even if `Time(["2000:001", np.nan])` could be handled, it just looks wrong to me. But I could be swayed on that.

**Comment by maxnoe** [comment]
```
Time(["2000:001", np.nan]) 
```
Here I completely agree, but there are many time formats which are represented by floats, for which nan is the most natural, of the top of my head `jd`, `mjd`, `unix`, `unix_tai`, probably more.

**Comment by taldcroft** [comment]
> or which nan is the most natural

If you mean for the purpose of indicating a missing value on **input**, I can certainly be convinced. This is in line with the principle of being liberal on input since other codes may well use `nan` to represent missing values for numerical time formats.

On **output**, there can only be one way of representing missing values, and that is `Masked`.

**Comment by mhvk** [comment]
Thanks, @taldcroft, for clarifying the discussion. Trying to follow up on that...

For a masked `float` one can distinguish between "missing" (masked)- and "messed up" (nan), where in normal usage "messed up" means that one tried to do an operation that is not possible, like dividing 0/0 or the log of a negative number.

I guess a primary question is whether it is useful in any way to distinguish between these cases in `Time`. If we decide it is not, i.e., we want `nan` and `NaT` just to signal something is missing *or* messed up, then we indeed need to be careful on input only, and if one wants `nan` one can convert masked elements on output (currently not super conveniently, but could be made better).

Distinguishing between missing and messed up has a real cost in code complexity, so the argument for it should be strong.

For `Time` itself, I think it is weak, since, there are no operations that mess it up. However, `TimeDelta` can be messed up and this can propagate:
```
t = Time("J2001")
dt = (t-t)/0
# /usr/lib/python3/dist-packages/astropy/time/utils.py:55: RuntimeWarning: invalid value encountered in scalar divide q1 = sum12 / divisor
dt
# <TimeDelta object: scale='tt' format='jd' value=nan>
```
At present, `Time` does not guard against such a `nan` entering:
```
t2 = t + dt
t2
# <Time object: scale='tt' format='jyear_str' value=Jnan>
t2.tai
# /usr/lib/python3/dist-packages/erfa/core.py:17280: RuntimeWarning: invalid value encountered in tttai
  tai1, tai2, c_retval = ufunc.tttai(tt1, tt2)
# <Time object: scale='tai' format='jyear_str' value=Jnan>
t2.isot
# /usr/lib/python3/dist-packages/erfa/core.py:16223: RuntimeWarning: invalid value encountered in d2dtf
  iy, im, id, ihmsf, c_retval = ufunc.d2dtf(scale, ndp, d1, d2)
# '-4713-11-24T-2147483648:-2147483648:-2147483648.-2147483648'
t2.datetime
# /usr/lib/python3/dist-packages/erfa/core.py:16223: RuntimeWarning: invalid value encountered in d2dtf
  iy, im, id, ihmsf, c_retval = ufunc.d2dtf(scale, ndp, d1, d2)
# ValueError: year must be in 1..9999, not -4713
```

**Comment by mhvk** [comment]
I've been thinking a bit more about implications, trying to keep in mind that whatever we do should not lead to a performance hit for the common case.

On input, I actually think the current behaviour is correct: if we are fed something like `Time([50000.0, np.nan], format="mjd")`, we should "refuse the temptation to guess"  and raise an error, not mask the element. It seems fine to me to leave it up to the user to pass in a `Masked` instance instead (they could use, e.g., `Time(np.ma.fix_invalid([50000.0, np.nan]), format="mjd")` -- which works as expected (we could document that, provide a similar function for `Masked`, or even introduce a new keyword argument).

For operations, I'd argue that instituting a post-operation check for `nan` inside an unmasked `Time` carries too high a cost for the common case to be worth it: let the warnings suffice. If the `Time` is masked already, then perhaps it is different. For `Masked` proper, it was a conscious choice not to do any checks, partially because of the performance hit but also because it is actually hard to do something beforehand that prevents the warnings (the checks in `np.ma.MaskedArray` are not great). But perhaps `Masked` itself, or the masked time, could have an option to convert any `NaN` to masked after each operation. That seems mostly a question of whether it is worth the performance hit and coding effort.

On output, I fear we do need to do something. For the masked case, I think just setting the mask on output is fine. For the unmasked case, changing to masked would seem surprising, and I think I'd prefer to just raise if the `NaN` cannot be represented (i.e., the non-float output formats). Getting `'-4713-11-24T-2147483648:-2147483648:-2147483648.-2147483648'` is really no good. And for string outputs, the perfomance hit will be minimal.

**Comment by taldcroft** [comment]
> For a masked float one can distinguish between "missing" (masked)- and "messed up" (nan), where in normal usage "messed up" means that one tried to do an operation that is not possible, like dividing 0/0 or the log of a negative number.

`nan` is commonly used for both "messed up" and "missing". The latter "missing" use is actually how I more frequently use it or encounter `nan`. For instance, Pandas missing data for float, or the FITS standard defines the use of `nan` for missing or invalid data for floats. Likewise Matplotlib essentially interprets `nan` as missing and I use this as a feature frequently in plotting.

So basically it is ambiguous which is why I'm not a huge fan of `nan`. But that doesn't mean we can't try to come up with some handling, it's just that defining something that makes everyone happy may be tricky.

What about saying that for astropy, `nan` is unambiguously "missing", meaning that it always converts to a masked item. That gives a simple story and would (I think) be relatively straightforward to implement in the current code framework.  (though I'm not sure about `dt = (t-t)/0`).
